access-grant/src/main/java/com/inrupt/client/accessgrant/AccessGrantClient.java

```diff
@@ -430,32 +430,28 @@ private <T extends AccessCredential> CompletionStage<List<T>> query(final URI re
             throw new AccessGrantException("Unsupported type " + clazz + " in query request");
         }
 
-        return v1Metadata().thenCompose(metadata -> {
-            final List<CompletableFuture<List<T>>> futures = buildQuery(config.getIssuer(), type,
-                    resource, creator, recipient, purposes, modes).stream()
-                .map(data -> Request.newBuilder(metadata.queryEndpoint)
-                        .header(CONTENT_TYPE, APPLICATION_JSON)
-                        .POST(Request.BodyPublishers.ofByteArray(serialize(data))).build())
-                .map(req -> client.send(req, Response.BodyHandlers.ofInputStream())
-                        .thenApply(res -> {
-                            try (final InputStream input = res.body()) {
-                                final int status = res.statusCode();
-                                if (isSuccess(status)) {
-                                    return processQueryResponse(input, supportedTypes, clazz);
-                                }
-                                throw new AccessGrantException("Unable to perform Access Grant query: HTTP error " +
-                                        status, status);
-                            } catch (final IOException ex) {
-                                throw new AccessGrantException(
-                                        "Unexpected I/O exception while processing Access Grant query", ex);
-                            }
-                        }).toCompletableFuture())
-                .collect(Collectors.toList());
-
-            return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
-                .thenApply(x -> futures.stream().map(CompletableFuture::join).flatMap(List::stream)
-                        .collect(Collectors.toList()));
-        });
+        return v1Metadata().thenApply(metadata -> buildQuery(config.getIssuer(), type,
```

Comment:
I'm not sure about the refactoring of this code. The use of CompletableFuture.allOf was intended to ensure that multiple requests were performed in parallel without being blocked by any particular response.

I know you mentioned that Jsonb objects are not thread safe, but that contradicts the contract stated in the javadocs:

    All the methods in this class are safe for use by multiple concurrent threads.

https://javaee.github.io/javaee-spec/javadocs/javax/json/bind/Jsonb.html